src/GafferOSL/OSLShader.cpp

```diff
@@ -131,6 +133,9 @@ ConstShadingEnginePtr getter( const ShadingEngineCacheGetterKey &key, size_t &co
 typedef IECorePreview::LRUCache<IECore::MurmurHash, ConstShadingEnginePtr, IECorePreview::LRUCachePolicy::Parallel, ShadingEngineCacheGetterKey> ShadingEngineCache;
 ShadingEngineCache g_shadingEngineCache( getter, 10000 );
 
+typedef boost::container::flat_set<IECore::InternedString> ShaderTypeSet;
+ShaderTypeSet g_compatibleShaders;
```

Comment:
This is likely going to fall foul of the "static initialization order fiasco" at some point. There is no guarantee that `g_compatibleShaders` will be initialised before any other static data (except other static data later in OSLShader.cpp), so we can end up with `static bool g_registration = registerCompatibleShader()` being done before the initialisation of `g_compatibleShaders`, leading to crashes. The way to avoid this is to package it into a function so that initialisation is guaranteed to occur on the first call :

```suggestion
ShaderTypeSet &compatibleShaders()
{
    ShaderTypeSet g_compatibleShaders;
    return g_compatibleShaders;
};
```

Reply:
Yeah I was wondering why you had this pattern, C++ is one esoteric beast.